x-pack/plugin/ent-search/src/main/java/org/elasticsearch/xpack/application/search/SearchApplicationTemplateService.java

```diff
@@ -0,0 +1,72 @@
+ * or more contributor license agreements. Licensed under the Elastic License
+ * 2.0; you may not use this file except in compliance with the Elastic License
+ * 2.0.
+ */
+
+package org.elasticsearch.xpack.application.search;
+
+import org.elasticsearch.common.ValidationException;
+import org.elasticsearch.common.xcontent.LoggingDeprecationHandler;
+import org.elasticsearch.script.Script;
+import org.elasticsearch.script.ScriptService;
+import org.elasticsearch.script.TemplateScript;
+import org.elasticsearch.search.builder.SearchSourceBuilder;
+import org.elasticsearch.xcontent.NamedXContentRegistry;
+import org.elasticsearch.xcontent.XContentFactory;
+import org.elasticsearch.xcontent.XContentParser;
+import org.elasticsearch.xcontent.XContentParserConfiguration;
+import org.elasticsearch.xcontent.XContentType;
+import org.elasticsearch.xpack.application.search.action.SearchApplicationSearchRequest;
+
+import java.io.IOException;
+import java.util.HashMap;
+import java.util.Map;
+
+public class SearchApplicationTemplateService {
+
+    private final ScriptService scriptService;
+    private final NamedXContentRegistry xContentRegistry;
+
+    public SearchApplicationTemplateService(ScriptService scriptService, NamedXContentRegistry xContentRegistry) {
+        this.scriptService = scriptService;
+        this.xContentRegistry = xContentRegistry;
+    }
+
+    public SearchSourceBuilder renderQuery(SearchApplication searchApplication, Map<String, Object> templateParams) throws IOException {
+        final Script script = searchApplication.searchApplicationTemplate().script();
+        TemplateScript compiledTemplate = scriptService.compile(script, TemplateScript.CONTEXT).newInstance(templateParams);
+        String requestSource = compiledTemplate.execute();
+        XContentParserConfiguration parserConfig = XContentParserConfiguration.EMPTY.withRegistry(xContentRegistry)
+            .withDeprecationHandler(LoggingDeprecationHandler.INSTANCE);
+        try (XContentParser parser = XContentFactory.xContent(XContentType.JSON).createParser(parserConfig, requestSource)) {
+            SearchSourceBuilder builder = SearchSourceBuilder.searchSource();
+            builder.parseXContent(parser, false);
+            return builder;
+        }
+    }
+
+    /**
+     * Renders the search application's associated template with the provided request parameters.
+     *
+     * @param searchApplication The SearchApplication we're accessing, to access the associated template
+     * @param request The search request, which is expected to contain a subset of the parameters included in the template.
+     * @return Map of all template parameters including template defaults for non-specified parameters
+     * @throws ValidationException on invalid template parameters
+     */
+    public Map<String, Object> renderTemplate(SearchApplication searchApplication, SearchApplicationSearchRequest request)
```

Comment:
I think it would be good to leave the Request on the Action level, and let the Service manage domain objects if possible instead of requests / responses.

Also, it makes me feel good to include possible exceptions (even Runtime ones) that can be thrown on the method:

```suggestion
    public Map<String, Object> renderTemplate(SearchApplication searchApplication, Map<String, Object> queryParams) throws ValidationException
```

Reply:
That's fair, I'll change 👍